subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -923,18 +923,16 @@ private ResolverResults resolveGraphForBuildDependenciesIfRequired() {
             return resolveToStateOrLater(GRAPH_RESOLVED).getCachedResolverResults();
         }
 
-        ResolveState currentState = currentResolveState.update(initial -> {
-            if (initial.state == UNRESOLVED) {
+        ResolveState currentState = currentResolveState.get();
+        if (currentState.state == UNRESOLVED) {
+            currentState = currentResolveState.update(initial -> {
```

Comment:
Shouldn´t the state condition be also kept in the `update {}` closure?

Reply:
Oh yeah. I missed the case where `currentState.state` could change after the first state condition and then before validating the lock. I fixed it.